docs/spec_draft.md

```diff
@@ -753,6 +754,183 @@ defined and one of the following:
+
+### Outputs
+
+| Name     | Type                                  |
+|----------|---------------------------------------|
+| `result` | tensor of any supported element types |
+
+### Constraints
+
+  * On Inputs
+    * (C1) rank(`operand`) $=$ size(`offset_dims`) $+$
+    size(`collapsed_slice_dims`)
+
+    * (C2) &nbsp; $0 \le$ `index_vector_dim` $\le$ rank(`start_indices`) Or
+    `index_vector_dim` $=$ rank(`start_indices`) in which case assume a trailing
+    $1$ on the shape of `start_indices`.
+
+    * (C3) dim(`start_indices`, `index_vector_dim`) $=$
+    size(`start_index_map`)
+
+    * (C4) All dimensions in `start_index_map` are unique.
+
+    * (C5) &nbsp; $0 \le$ `start_index_map`[i] $\lt$ rank(`operand`) $\forall i$
+    such that $0 \le$ i $\lt$ size(`start_index_map`)
+
+    * (C6) size(`slice_sizes`) $=$ rank(`operand`).
+
+    * (C7) &nbsp; $0 \le$ `slice_sizes`[i] $\lt$ shape(`operand`)[i] $\forall
+    i$ such that $0 \le$ i $\lt$ size(`slice_sizes`)
+
+    * (C8) All dimensions in `offset_dims` are unique and sorted.
+
+    * (C9) &nbsp; $0 \le$ `offset_dims`[i] $\lt$ rank(`result`) $\forall i$
+    such that $0 \le$ i $\lt$ size(`offset_dims`)
+
+    * (C10) All dimensions in `collapsed_slice_dims` are unique and sorted.
+
+    * (C11) &nbsp; $0 \le$ `collapsed_slice_dims`[i] $\lt$ size(`slice_sizes`)
+    $\forall i$ such that $0 \le$ i $\lt$ size(`collapsed_slice_dims`)
+
+    * (C12) `slize_sizes`[i] $\le$ 1 $\forall i \in$ `collapsed_slice_dims`
+
+  * On shape of `result`
+    * rank(`result`) $=$ size(`offset_dims`) $+$ rank(`start_indices`) - 1
+    * The dimension size of `i`$^{th}$ dimension, `di`, of `result` is computed
+      as:
+      * If `i` is present in the `batch_dims`, i.e., `i` $=$ `batch_dims`[`k`]
+      for some `k`, then
+        * `di` $=$ dim(`start_indices`, `k`) if `k` $\lt$ `index_vector_dim`
+        * `di` $=$ dim(`start_indices`, `k+1`) otherwise.
+      * If `i` is present in `offset_dims`, i.e., `i` $=$ `offset_dims`[`k`] for
+      some `k`, then `di` $=$ `adjusted_slice_sizes`[`k`] where
+      `adjusted_slice_sizes`= `slize_sizes` - {`size_sizes[i]` : i $\in$
+      `collapsed_slice_dims`}.
+
+### Examples
+
+```mlir
+  // %operand: [
+  //            [[1, 2], [3, 4], [5, 6], [7, 8]],
+  //            [[9, 10],[11, 12], [13, 14], [15, 16]],
+  //            [[17, 18], [19, 20], [21, 22], [23, 24]]
+  //           ]
+  // %start_indices: [
+  //                  [[0, 0], [1, 0], [2, 1]],
+  //                  [[0, 1], [1, 1], [0, 2]]]
+  //                 ]
+  %result = "stablehlo.gather"(%operand, %start_indices) {
+    dimension_numbers = #stablehlo.gather<
+      offset_dims = [2, 3],
+      collapsed_slice_dims = [0],
+      start_index_map = [0, 2],
+      index_vector_dim = 2,
+    >,
+    slice_sizes = dense<[0, 2, 2]> : tensor<3xi64>,
+    indices_are_sorted = false
```

Comment:
Let's use exactly the same inputs as in the diagram? This example is slightly different. (Same below).

```
  %result = "stablehlo.gather"(%operand, %start_indices) {
    dimension_numbers = #stablehlo.gather<
      offset_dims = [2, 3],
      collapsed_slice_dims = [0],
      start_index_map = [1, 0],
      index_vector_dim = 2,
    >,
    slice_sizes = dense<[1, 2, 2]> : tensor<3xi64>,
    indices_are_sorted = false
  } : (tensor<3x4x2xi32>, tensor<2x3x2xi64>) -> tensor<2x3x2x2xi32>
```

Interestingly, the `%result` below is incorrect for the example in the pull request, but is correct for the example from the diagram.

Reply:
In the example, I was doing the computation assuming the same start_index_map in the image, but wrote a different one.
fixed.